Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift

```
@@ -0,0 +1,377 @@
+            self.paymentFlowClientDelegate?.onPaymentComplete(result, error: error)
+        }
+    }
+    
+    private func logThreeDSecureCompletedAnalytics(forResult result: BTThreeDSecureResult, apiClient: BTAPIClient) {
+        let liabilityShiftPossible = result.tokenizedCard?.threeDSecureInfo.liabilityShiftPossible ?? false
+        apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.liability-shift-possible.\(stringFor(liabilityShiftPossible))")
+
+        let liabilityShifted = result.tokenizedCard?.threeDSecureInfo.liabilityShiftPossible ?? false
+        apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.liability-shifted.\(liabilityShifted)")
+
+        apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.completed")
+    }
+    
+    func stringFor(_ boolean: Bool) -> String {
+        boolean ? "true" : "false"
+    }
+}
+
+// MARK: - BTPaymentFlowRequestDelegate Protocol Conformance
+
+extension BTThreeDSecureRequest: BTPaymentFlowRequestDelegate {
+
+    /// :nodoc:
+    public func handle(
+        _ request: BTPaymentFlowRequest,
+        client apiClient: BTAPIClient,
+        paymentClientDelegate delegate: BTPaymentFlowClientDelegate
+    ) {
+        paymentFlowClientDelegate = delegate
+
+        apiClient.sendAnalyticsEvent("ios.three-d-secure.initialized")
+
+        apiClient.fetchOrReturnRemoteConfiguration { [weak self] configuration, error in
+            guard let self else { return }
+
+            if let error {
+                self.paymentFlowClientDelegate?.onPaymentComplete(nil, error: error)
+                return
+            }
+
+            var integrationError: Error?
+
+            if configuration?.cardinalAuthenticationJWT == nil {
+                NSLog("%@ BTThreeDSecureRequest versionRequested is 2, but merchant account is not setup properly.", BTLogLevelDescription.string(for: .critical))
```

So we are not dealing with version 1 anymore, right?

---

Not in v6, no - v5 will need to return an error for merchants still attempting to use 3DSv1. All v1 classes were removed here: https://github.com/braintree/braintree_ios/pull/979